### DISCUSSION TITLE
Serious secuirty bridge in manual cluster

### DISCUSSION THREAD

**Discussion by roysG** [emqx/emqx#7235] [Q&A]
Hi,
As describe in the docs:
https://www.emqx.io/docs/en/v4.4/getting-started/cluster.html#node-discovery-and-autocluster

In case, attacker download and install emqx in his server, 
And in case he know the hostname and the ip/domain name of one of the joined servers, he can joined to the clustering with the command:

"emqx_ctl cluster join emqx@s1.emqx.io" 

After that the attacker have full control on the cluster and he can tell all others servers in the clustering to force leave, with the command:
"emqx_ctl cluster force-leave emqx@s2.emqx.io"

And the cluster of the domain will be down!!!

This is a serious hack. Do i miss something?

In case this is a security bridge, 
Do we have the ability to allow this commands just for specifics ips to send request of join/force-leave?

**Comment by zmstone** [comment]
Hi @roysG 

There are few ways to mitigate it.

* One should never expose ports other than the `external` listener ports to public access.
  Sensitive ports are
  - Erlang distribution `4370+N `, `N` being the number suffix for each node 
  - Intra cluster RPC `5370+N`, `N` being the number suffix for each node
* Use a secret value for `node.cookie`

**Reply by roysG** [reply]
This link answer and help me to prevent it!
https://github.com/emqx/emqx-tutorial/blob/master/en/cluster/setup_start.md
Thanks you.

**Reply by roysG** [reply]
1. In case i added strong node.cookies, the port 4370 and 5370 may be opened externally without any worry, right?
2. In case i want also to block those ports, can you explain me further about the +N?, i did not not understand it
3. in case i am using static option, do i have to insert all those ips in each server?
4. I considering to using in mcast or etcd server, but i could not found any example on the internet. can you give me or direct me to instruction? Also tell me what is your recommendation? Thanks.

**Reply by zmstone** [reply]
1. Cookie can be compromised, always keep the ports internal.
2. If the node name suffixed without number, e.g. the default name `emqx@127.0.0.1` then offset is 0, for `node1@host.name` the offset is 1
3. All nodes in a emqx cluster should have static name, number suffix before the `@` is usually used when starting more than one emqx node in one host (mostly testing env)
4. We test with etcd a lot.
Doc here: https://www.emqx.io/docs/en/v4.4/advanced/cluster.html#node-discovery-and-automatic-clustering The doc is rather brief, if you need more info, we can improve it

**Comment by roysG** [comment]
Can you send me further installation and your configuration file for etcd?
Thanks

**Reply by zmstone** [reply]
You can maybe take this as ref: https://github.com/emqx/cdk-emqx-cluster/blob/c120196331b6e013760e47f51938309f93760e41/cdk_emqx_cluster/cdk_emqx_cluster_stack.py#L677

**Reply by roysG** [reply]
Thanks
